### PR TITLE
Generic fixtures for parametrize usage

### PR DESCRIPTION
**Pull request by RazTamir**
- Added a fixture to iterate over ceph interfaces
 - Created generic fixtures to create resources based on interface

### CODE REVIEW

**Comment by RazTamir** [comment]
the plan is to replace the calls to these fixtures PR by PR to make it more smooth conversion

**Comment by clacroix12** [comment]
Why are we adding more fixtures to `fixtures.py` which use class instance references when we have already decided as a team to move away from this approach? If additional fixtures are needed why are we not implementing them in the way our documentation describes?

**Comment by RazTamir** [comment]
Verified:
http://pastebin.test.redhat.com/787801

File: tests/conftest.py [hunk 1]
```diff
@@ -578,3 +555,18 @@ def run_io_in_bg():
 
         thread = threading.Thread(target=run_io_in_bg)
         thread.start()
+
+
+@pytest.fixture(
```
Comment (ebenahar) [inline tests/conftest.py]: does it mean every test will have to run on both cephfs and rbd?

Comment (RazTamir) [reply tests/conftest.py]: no 'have to' but optional

File: tests/conftest.py [hunk 2]
```diff
@@ -338,14 +315,15 @@ def factory(
...
+        status=constants.STATUS_RUNNING,
     ):
         """
         Args:
             interface (str): CephBlockPool or CephFileSystem. This decides
                 whether a RBD based or CephFS resource is created.
                 RBD is default.
-            pvc (object): ocs_ci.ocs.resources.ocs.OCS instance of 'PVC' kind.
+            pvc (PVC object): ocs_ci.ocs.resources.pvc.PVC instance
+            kind.
```
Comment (ebenahar) [inline tests/conftest.py]: missing indentation

File: tests/e2e/test_io_performance.py [hunk 3]
```diff
@@ -4,35 +4,27 @@
...
-    create_rbd_pod.__name__,
-)
 @google_api_required
 @tier1
 class TestIOPerformance(ManageTest):
     """
     Test IO performance
     """
+    @pytest.fixture()
+    def base_setup(self, request, interface_iterate, pod_factory):
```
Comment (ebenahar) [inline tests/e2e/test_io_performance.py]: missing docstring

File: tests/conftest.py [hunk 4]
```diff
@@ -225,14 +214,13 @@ def finalizer():
...
-                    resource_name=instance.namespace
-                )
-                instance.ocp.wait_for_delete(
-                    instance.namespace
-                )
+            ocp.switch_to_default_rook_cluster_project()
+            instance.delete(
+                resource_name=instance.namespace
+            )
+            instance.wait_for_delete(
```
Comment (ebenahar) [inline tests/conftest.py]: should fit in 1 line

File: tests/conftest.py [hunk 5]
```diff
@@ -84,18 +84,24 @@ def finalizer():
...
-    def factory():
-        cbp_obj = helpers.create_ceph_block_pool()
-        assert cbp_obj, "Failed to create block pool"
-        instances.append(cbp_obj)
-        return cbp_obj
+    def factory(interface=constants.CEPHBLOCKPOOL):
+        if interface == constants.CEPHBLOCKPOOL:
+            ceph_pool_obj = helpers.create_ceph_block_pool()
+        elif interface == constants.CEPHFILESYSTEM:
+            cfs = ocp.OCP(kind=constants.CEPHFILESYSTEM).get(
```
Comment (ebenahar) [inline tests/conftest.py]: I think you need to initialize OCP with openshift-storage namespace

Comment (RazTamir) [reply tests/conftest.py]: Not mandatory as the ceph pool is namespace-less

File: tests/conftest.py (hunk 4 above)
Comment (RazTamir) [reply tests/conftest.py]: done

File: tests/conftest.py (hunk 2 above)
Comment (RazTamir) [reply tests/conftest.py]: done

File: tests/e2e/test_io_performance.py (hunk 3 above)
Comment (RazTamir) [reply tests/e2e/test_io_performance.py]: done

File: tests/conftest.py [hunk 6]
```diff
@@ -152,7 +159,7 @@ def factory(
             secret = secret or secret_factory(interface=interface)
 
             if interface == constants.CEPHBLOCKPOOL:
```
Comment (fbalak) [inline tests/conftest.py]: This way it adds custom pool only to RBD. If you want it for CephFS too, you should remove interface condition and use something like:
```
block_pool = block_pool or ceph_pool_factory(interface=interface)
interface_name = block_pool.name
```
Also block_pool variable should be renamed to pool.

File: ocs_ci/utility/environment_check.py [hunk 7]
```diff
@@ -17,18 +17,16 @@
 CEPHBLOCKPOOL = ocp.OCP(kind=constants.CEPHBLOCKPOOL)
 PV = ocp.OCP(kind=constants.PV)
 PVC = ocp.OCP(kind=constants.PVC)
-SECRET = ocp.OCP(kind=constants.SECRET)
```
[the remark this answers is not in the document]
Comment (RazTamir) [reply ocs_ci/utility/environment_check.py]: That's from another PR.
Not sure why it is here as well

File: tests/conftest.py [hunk 8]
```diff
@@ -84,18 +84,25 @@ def finalizer():
 
 
 @pytest.fixture()
-def block_pool_factory(request):
+def ceph_pool_factory(request):
     """
     Create a Ceph block pool factory.
     Calling this fixture creates new block pool instance.
```
[the remark this answers is not in the document]
Comment (RazTamir) [reply tests/conftest.py]: done

File: tests/conftest.py (hunk 6 above)
Comment (RazTamir) [reply tests/conftest.py]: done

File: tests/conftest.py [hunk 9]
```diff
@@ -84,18 +84,25 @@ def finalizer():
...
-        cbp_obj = helpers.create_ceph_block_pool()
-        assert cbp_obj, "Failed to create block pool"
-        instances.append(cbp_obj)
-        return cbp_obj
+    def factory(interface=constants.CEPHBLOCKPOOL):
+        if interface == constants.CEPHBLOCKPOOL:
+            ceph_pool_obj = helpers.create_ceph_block_pool()
+        elif interface == constants.CEPHFILESYSTEM:
+            cfs = ocp.OCP(
+                kind=constants.CEPHFILESYSTEM,
```
[the remark this answers is not in the document]
Comment (RazTamir) [reply tests/conftest.py]: yes that's right

File: ocs_ci/utility/environment_check.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (fbalak) [reply ocs_ci/utility/environment_check.py]: So it should be probably removed from this PR...

**Comment by vasukulkarni** [comment]
@RazTamir needs rebase to remove the secret PR.

File: tests/helpers.py [hunk 10]
```diff
@@ -225,6 +225,32 @@ def create_ceph_block_pool(pool_name=None):
...
+        OCS: An OCS instance for the Ceph file system
+    """
+    cfs_data = templating.load_yaml_to_dict(constants.CEPHFILESYSTEM_YAML)
+    cfs_data['metadata']['name'] = (
+        pool_name if pool_name else create_unique_resource_name(
+            'test', 'cfs'
+        )
+    )
+    cfs_data['metadata']['namespace'] = defaults.ROOK_CLUSTER_NAMESPACE
+    cfs_data = create_resource(**cfs_data, wait=False)
```
Comment (vasukulkarni) [inline tests/helpers.py]: s/cfs_data/cfs_obj

File: tests/conftest.py (hunk 1 above)
Comment (vasukulkarni) [reply tests/conftest.py]: I find this part of code confusing? doesn't it iterate for params,  can you explain in docstring here?

File: ocs_ci/utility/environment_check.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (vasukulkarni) [reply ocs_ci/utility/environment_check.py]: Once rebase is done this should go away

**Comment by vasukulkarni** [comment]
few comments, otherwise LGTM.

**Comment by RazTamir** [comment]
The changes requested by @fbalak are now resolved

**Pull request merged**